File: mcp-server/mentraflow-mcp-server.js
#!/usr/bin/env node

/**
 * MentraFlow MCP Server
 * Bridges Perplexity/Claude to MentraFlow API
 * Protocol: JSON-RPC over stdio
 */

const https = require('https');

// MentraFlow API endpoint (from environment or default)
const MENTRAFLOW_API = process.env.MENTRAFLOW_ENDPOINT || 'https://resume-session-13.preview.emergentagent.com/api/mcp/receive-export';
const DEFAULT_USER_ID = process.env.MENTRAFLOW_USER_ID || null;

// JSON-RPC response helper
function sendResponse(id, result) {
  const response = {
    jsonrpc: '2.0',
    id: id,
    result: result
  };
  console.log(JSON.stringify(response));
}

function sendError(id, code, message) {
  const response = {
    jsonrpc: '2.0',
    id: id,
    error: {
      code: code,
      message: message
    }
  };
  console.log(JSON.stringify(response));
}

// Send data to MentraFlow
function exportToMentraFlow(userId, conversations) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify({
      user_id: userId,
      conversations: conversations,
      export_timestamp: new Date().toISOString()
    });

    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': data.length
      }
    };

    const req = https.request(MENTRAFLOW_API, options, (res) => {
      let body = '';
      res.on('data', (chunk) => body += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) {
          resolve(JSON.parse(body));
        } else {
          reject(new Error(`API returned ${res.statusCode}: ${body}`));
        }
      });
    });

    req.on('error', (e) => reject(e));
    req.write(data);
    req.end();
  });
}

// Handle MCP protocol messages
function handleMessage(message) {
  try {
    const request = JSON.parse(message);
    
    // Handle different MCP methods
    switch (request.method) {
      case 'initialize':
        sendResponse(request.id, {
          protocolVersion: '2024-11-05',
          serverInfo: {
            name: 'mentraflow',
            version: '1.0.0'
          },
          capabilities: {
            tools: {}
          }
        });
        break;

      case 'tools/list':
        sendResponse(request.id, {
          tools: [
            {
              name: 'export_to_mentraflow',
              description: 'Export conversations to MentraFlow for knowledge retention and quiz generation',
              inputSchema: {
                type: 'object',
                properties: {
                  user_id: {
                    type: 'string',
                    description: 'MentraFlow user ID'
                  },
                  conversations: {
                    type: 'array',
                    description: 'Array of conversations to export',
                    items: {
                      type: 'object',
                      properties: {
                        conversation_id: { type: 'string' },
                        platform: { type: 'string' },
                        title: { type: 'string' },
                        messages: {
                          type: 'array',
                          items: {
                            type: 'object',
                            properties: {
                              role: { type: 'string' },
                              content: { type: 'string' }
                            }
                          }
                        }
                      }
                    }
                  }
                },
                required: ['user_id', 'conversations']
              }
            }
          ]
        });
        break;

      case 'tools/call':
        if (request.params.name === 'export_to_mentraflow') {
          const { user_id, conversations } = request.params.arguments;
          
          exportToMentraFlow(user_id, conversations)
            .then((result) => {
              sendResponse(request.id, {
                content: [
                  {
                    type: 'text',
                    text: `✅ Successfully exported ${conversations.length} conversation(s) to MentraFlow!\n\n` +
                          `📊 Processing Summary:\n` +
                          `- Concepts extracted: ${result.concepts_extracted || 'Processing...'}\n` +
                          `- Quizzes generated: ${result.quizzes_generated || 'Processing...'}\n` +
                          `- Import ID: ${result.import_id}\n\n` +
                          `Your knowledge graph is being updated. Check MentraFlow dashboard for new concepts and quizzes!`
                  }
                ]
              });
            })
            .catch((error) => {
              sendError(request.id, -32603, `Failed to export: ${error.message}`);
            });
        } else {
          sendError(request.id, -32601, 'Method not found');
        }
        break;

      default:
        sendError(request.id, -32601, 'Method not found');
    }
  } catch (error) {
    console.error('Error handling message:', error);
    sendError(null, -32700, 'Parse error');
  }
}

// Main: Read from stdin, write to stdout
let buffer = '';

process.stdin.on('data', (chunk) => {
  buffer += chunk.toString();
  
  // Process complete messages (newline-delimited JSON)
  const lines = buffer.split('\n');
  buffer = lines.pop() || ''; // Keep incomplete line in buffer
  
  lines.forEach((line) => {
    if (line.trim()) {
      handleMessage(line);
    }
  });
});

process.stdin.on('end', () => {
  process.exit(0);
});

// Log startup to stderr (doesn't interfere with JSON-RPC on stdout)
console.error('MentraFlow MCP Server started');
console.error(`API Endpoint: ${MENTRAFLOW_API}`);

File: mcp-server/mentraflow-mcp-server.py
#!/usr/bin/env python3
"""
MentraFlow MCP Server (Python)
Bridges Perplexity/Claude to MentraFlow API
Protocol: JSON-RPC over stdio
"""

import json
import sys
import requests
from datetime import datetime

MENTRAFLOW_API = 'https://resume-session-13.preview.emergentagent.com/api/mcp/receive-export'

def send_response(request_id, result):
    """Send JSON-RPC success response"""
    response = {
        'jsonrpc': '2.0',
        'id': request_id,
        'result': result
    }
    print(json.dumps(response), flush=True)

def send_error(request_id, code, message):
    """Send JSON-RPC error response"""
    response = {
        'jsonrpc': '2.0',
        'id': request_id,
        'error': {
            'code': code,
            'message': message
        }
    }
    print(json.dumps(response), flush=True)

def export_to_mentraflow(user_id, conversations):
    """Send data to MentraFlow API"""
    try:
        payload = {
            'user_id': user_id,
            'conversations': conversations,
            'export_timestamp': datetime.utcnow().isoformat()
        }
        
        response = requests.post(
            MENTRAFLOW_API,
            json=payload,
            headers={'Content-Type': 'application/json'},
            timeout=30
        )
        
        response.raise_for_status()
        return response.json()
    except Exception as e:
        raise Exception(f"Failed to export: {str(e)}")

def handle_message(message):
    """Handle incoming MCP protocol message"""
    try:
        request = json.loads(message)
        request_id = request.get('id')
        method = request.get('method')
        
        if method == 'initialize':
            send_response(request_id, {
                'protocolVersion': '2024-11-05',
                'serverInfo': {
                    'name': 'mentraflow',
                    'version': '1.0.0'
                },
                'capabilities': {
                    'tools': {}
                }
            })
        
        elif method == 'tools/list':
            send_response(request_id, {
                'tools': [
                    {
                        'name': 'export_to_mentraflow',
                        'description': 'Export conversations to MentraFlow for knowledge retention and quiz generation',
                        'inputSchema': {
                            'type': 'object',
                            'properties': {
                                'user_id': {
                                    'type': 'string',
                                    'description': 'MentraFlow user ID'
                                },
                                'conversations': {
                                    'type': 'array',
                                    'description': 'Array of conversations to export'
                                }
                            },
                            'required': ['user_id', 'conversations']
                        }
                    }
                ]
            })
        
        elif method == 'tools/call':
            params = request.get('params', {})
            if params.get('name') == 'export_to_mentraflow':
                args = params.get('arguments', {})
                user_id = args.get('user_id')
                conversations = args.get('conversations', [])
                
                result = export_to_mentraflow(user_id, conversations)
                
                send_response(request_id, {
                    'content': [
                        {
                            'type': 'text',
                            'text': f"✅ Successfully exported {len(conversations)} conversation(s) to MentraFlow!\n\n"
                                   f"📊 Processing Summary:\n"
                                   f"- Import ID: {result.get('import_id', 'N/A')}\n"
                                   f"- Status: {result.get('message', 'Processing...')}\n\n"
                                   f"Your knowledge graph is being updated. Check MentraFlow dashboard!"
                        }
                    ]
                })
            else:
                send_error(request_id, -32601, 'Method not found')
        
        else:
            send_error(request_id, -32601, 'Method not found')
    
    except json.JSONDecodeError:
        send_error(None, -32700, 'Parse error')
    except Exception as e:
        send_error(request_id if 'request' in locals() else None, -32603, str(e))

def main():
    """Main loop: read from stdin, write to stdout"""
    print('MentraFlow MCP Server started', file=sys.stderr)
    print(f'API Endpoint: {MENTRAFLOW_API}', file=sys.stderr)
    
    for line in sys.stdin:
        line = line.strip()
        if line:
            handle_message(line)

if __name__ == '__main__':
    main()
